Skywalking agent fails to establish gRPC channel with TLS when no CA files

Please answer these questions before submitting your issue.

- Why do you submit this issue?
  Skywallking agent fails to establish gRPC channel with TLS except CA file exists. Actually the CA file is not required by just initial a TLS channel instead of a plian text channel. 

- [ ] Question or discussion
- [ ] Bug
- [ ] Requirement
   The skywalking agent can establish a TLS channel talking with gRPC without CA file.
   
- [ ] Feature or performance improvement

___
### Question
- What do you want to know?

___
### Bug
- Which version of SkyWalking, OS, and JRE?
   8.3.0
- Which company or project?

- What happened?
If possible, provide a way to reproduce the error. e.g. demo application, component version.

___
### Requirement or improvement
- Please describe your requirements or improvement suggestions.
  Skywalking agent can establish a gRPC tls channel without CA file by just setting the negotiation type to TLS instead of plain text.

The above change to TLSChannelBuilder.java can establish a tls grpc channel without CA file. Currently, the grpc channel failed to established due to following exceptions:

ERROR 2020-12-21 15:37:07:256 SkywalkingAgent-6-JVMService-consume-0 JVMMetricsSender : send JVM metrics to Collector fail. 
org.apache.skywalking.apm.dependencies.io.grpc.StatusRuntimeException: INTERNAL: http2 exception
	at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:262)
	at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:243)
	at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:156)
	at org.apache.skywalking.apm.network.language.agent.v3.JVMMetricReportServiceGrpc$JVMMetricReportServiceBlockingStub.collect(JVMMetricReportServiceGrpc.java:181)
	at org.apache.skywalking.apm.agent.core.jvm.JVMMetricsSender.run(JVMMetricsSender.java:82)
	at org.apache.skywalking.apm.util.RunnableWithExceptionProtection.run(RunnableWithExceptionProtection.java:33)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.skywalking.apm.dependencies.io.netty.handler.codec.http2.Http2Exception: First received frame was not SETTINGS. Hex dump for first 5 bytes: 485454502f
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.http2.Http2Exception.connectionError(Http2Exception.java:103)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.http2.Http2ConnectionHandler$PrefaceDecoder.verifyFirstFrameIsSettings(Http2ConnectionHandler.java:338)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.http2.Http2ConnectionHandler$PrefaceDecoder.decode(Http2ConnectionHandler.java:239)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:438)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	at org.apache.skywalking.apm.dependencies.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at org.apache.skywalking.apm.dependencies.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at org.apache.skywalking.apm.dependencies.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at org.apache.skywalking.apm.dependencies.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at org.apache.skywalking.apm.dependencies.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -21,6 +21,7 @@ Release Notes.
 * Fix thrift plugin collects wrong args when the method without parameter.
 * Fix DataCarrier's `org.apache.skywalking.apm.commons.datacarrier.buffer.Buffer` implementation isn't activated in `IF_POSSIBLE` mode.
 * Fix ArrayBlockingQueueBuffer's useless `IF_POSSIBLE` mode list
+* Support building gRPC TLS channel but CA file is not required.
 
 #### OAP-Backend
 * Make meter receiver support MAL.

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
@@ -124,6 +124,11 @@ public class Config {
          * Keep tracing even the backend is not available.
          */
         public static boolean KEEP_TRACING = false;
+
+        /**
+         * Should TLS enabled for gRPC channels
+         */
+        public static boolean IS_GRPC_CHANNEL_TLS_FORCED = true;
     }
 
     public static class OsInfo {

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
@@ -126,7 +126,7 @@ public class Config {
         public static boolean KEEP_TRACING = false;
 
         /**
-         * Should TLS enabled for gRPC channels
+         * Force open TLS for gRPC channel if true.
          */
         public static boolean FORCE_TLS = false;
     }

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
@@ -128,7 +128,7 @@ public class Config {
         /**
          * Should TLS enabled for gRPC channels
          */
-        public static boolean IS_GRPC_CHANNEL_TLS_FORCED = true;
+        public static boolean FORCE_TLS = true;
     }
 
     public static class OsInfo {

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java
@@ -128,7 +128,7 @@ public class Config {
         /**
          * Should TLS enabled for gRPC channels
          */
-        public static boolean FORCE_TLS = true;
+        public static boolean FORCE_TLS = false;
     }
 
     public static class OsInfo {

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
@@ -26,6 +26,7 @@ import java.io.File;
 import javax.net.ssl.SSLException;
 import org.apache.skywalking.apm.agent.core.boot.AgentPackageNotFoundException;
 import org.apache.skywalking.apm.agent.core.boot.AgentPackagePath;
+import org.apache.skywalking.apm.agent.core.conf.Config;
 import org.apache.skywalking.apm.agent.core.conf.Constants;
 
 /**
@@ -37,12 +38,14 @@ public class TLSChannelBuilder implements ChannelBuilder<NettyChannelBuilder> {
     @Override
     public NettyChannelBuilder build(
         NettyChannelBuilder managedChannelBuilder) throws AgentPackageNotFoundException, SSLException {
-        File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
-        if (caFile.exists() && caFile.isFile()) {
+        if (Config.Agent.IS_GRPC_CHANNEL_TLS_FORCED) {
             SslContextBuilder builder = GrpcSslContexts.forClient();
-            builder.trustManager(caFile);
+            File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
+            if (caFile.exists() && caFile.isFile()) {
+                builder.trustManager(caFile);
+            }
             managedChannelBuilder = managedChannelBuilder.negotiationType(NegotiationType.TLS)
-                                                         .sslContext(builder.build());
+                    .sslContext(builder.build());
         }
         return managedChannelBuilder;
     }

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
@@ -45,7 +45,7 @@ public class TLSChannelBuilder implements ChannelBuilder<NettyChannelBuilder> {
                 builder.trustManager(caFile);
             }
             managedChannelBuilder = managedChannelBuilder.negotiationType(NegotiationType.TLS)
-                    .sslContext(builder.build());
+                                                         .sslContext(builder.build());
         }
         return managedChannelBuilder;
     }

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
@@ -38,10 +38,11 @@ public class TLSChannelBuilder implements ChannelBuilder<NettyChannelBuilder> {
     @Override
     public NettyChannelBuilder build(
         NettyChannelBuilder managedChannelBuilder) throws AgentPackageNotFoundException, SSLException {
-        if (Config.Agent.FORCE_TLS) {
+        File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
+	boolean isCAFileExist = caFile.exists() && caFile.isFile();
+        if (Config.Agent.FORCE_TLS || isCAFileExist) {
             SslContextBuilder builder = GrpcSslContexts.forClient();
-            File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
-            if (caFile.exists() && caFile.isFile()) {
+            if (isCAFileExist) {
                 builder.trustManager(caFile);
             }
             managedChannelBuilder = managedChannelBuilder.negotiationType(NegotiationType.TLS)

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
@@ -38,7 +38,7 @@ public class TLSChannelBuilder implements ChannelBuilder<NettyChannelBuilder> {
     @Override
     public NettyChannelBuilder build(
         NettyChannelBuilder managedChannelBuilder) throws AgentPackageNotFoundException, SSLException {
-        if (Config.Agent.IS_GRPC_CHANNEL_TLS_FORCED) {
+        if (Config.Agent.FORCE_TLS) {
             SslContextBuilder builder = GrpcSslContexts.forClient();
             File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
             if (caFile.exists() && caFile.isFile()) {

--- a/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
+++ b/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TLSChannelBuilder.java
@@ -39,7 +39,7 @@ public class TLSChannelBuilder implements ChannelBuilder<NettyChannelBuilder> {
     public NettyChannelBuilder build(
         NettyChannelBuilder managedChannelBuilder) throws AgentPackageNotFoundException, SSLException {
         File caFile = new File(AgentPackagePath.getPath(), CA_FILE_NAME);
-	boolean isCAFileExist = caFile.exists() && caFile.isFile();
+        boolean isCAFileExist = caFile.exists() && caFile.isFile();
         if (Config.Agent.FORCE_TLS || isCAFileExist) {
             SslContextBuilder builder = GrpcSslContexts.forClient();
             if (isCAFileExist) {

--- a/apm-sniffer/config/agent.config
+++ b/apm-sniffer/config/agent.config
@@ -51,8 +51,8 @@ agent.service_name=${SW_AGENT_NAME:Your_ApplicationName}
 # Notice, in the current practice, we don't recommend the length over 190.
 # agent.operation_name_threshold=${SW_AGENT_OPERATION_NAME_THRESHOLD:150}
 
-# If true, SkyWalking agent will force to build a TLS gRPC channel, and construct trust manager when file `/ca/ca.crt` exists,
-# otherwise a plain text gRPC channel to be built.
+# The agent use gRPC plain text in default.
+# If true, SkyWalking agent uses TLS even no CA file detect.
 # agent.force_tls=${SW_AGENT_FORCE_TLS:false}
 
 # If true, skywalking agent will enable profile when user create a new profile task. Otherwise disable profile.

--- a/apm-sniffer/config/agent.config
+++ b/apm-sniffer/config/agent.config
@@ -53,7 +53,7 @@ agent.service_name=${SW_AGENT_NAME:Your_ApplicationName}
 
 # If true, SkyWalking agent will force to build a TLS gRPC channel, and construct trust manager when file `/ca/ca.crt` exists,
 # otherwise a plain text gRPC channel to be built.
-# agent.is_grpc_channel_tls_forced=${SW_AGENT_IS_GRPC_CHANNEL_TLS_FORCED:true}
+# agent.force_tls=${SW_AGENT_FORCE_TLS:true}
 
 # If true, skywalking agent will enable profile when user create a new profile task. Otherwise disable profile.
 # profile.active=${SW_AGENT_PROFILE_ACTIVE:true}

--- a/apm-sniffer/config/agent.config
+++ b/apm-sniffer/config/agent.config
@@ -51,6 +51,10 @@ agent.service_name=${SW_AGENT_NAME:Your_ApplicationName}
 # Notice, in the current practice, we don't recommend the length over 190.
 # agent.operation_name_threshold=${SW_AGENT_OPERATION_NAME_THRESHOLD:150}
 
+# If true, SkyWalking agent will force to build a TLS gRPC channel, and construct trust manager when file `/ca/ca.crt` exists,
+# otherwise a plain text gRPC channel to be built.
+# agent.is_grpc_channel_tls_forced=${SW_AGENT_IS_GRPC_CHANNEL_TLS_FORCED:true}
+
 # If true, skywalking agent will enable profile when user create a new profile task. Otherwise disable profile.
 # profile.active=${SW_AGENT_PROFILE_ACTIVE:true}
 

--- a/apm-sniffer/config/agent.config
+++ b/apm-sniffer/config/agent.config
@@ -52,7 +52,7 @@ agent.service_name=${SW_AGENT_NAME:Your_ApplicationName}
 # agent.operation_name_threshold=${SW_AGENT_OPERATION_NAME_THRESHOLD:150}
 
 # The agent use gRPC plain text in default.
-# If true, SkyWalking agent uses TLS even no CA file detect.
+# If true, SkyWalking agent uses TLS even no CA file detected.
 # agent.force_tls=${SW_AGENT_FORCE_TLS:false}
 
 # If true, skywalking agent will enable profile when user create a new profile task. Otherwise disable profile.

--- a/apm-sniffer/config/agent.config
+++ b/apm-sniffer/config/agent.config
@@ -53,7 +53,7 @@ agent.service_name=${SW_AGENT_NAME:Your_ApplicationName}
 
 # If true, SkyWalking agent will force to build a TLS gRPC channel, and construct trust manager when file `/ca/ca.crt` exists,
 # otherwise a plain text gRPC channel to be built.
-# agent.force_tls=${SW_AGENT_FORCE_TLS:true}
+# agent.force_tls=${SW_AGENT_FORCE_TLS:false}
 
 # If true, skywalking agent will enable profile when user create a new profile task. Otherwise disable profile.
 # profile.active=${SW_AGENT_PROFILE_ACTIVE:true}

--- a/docs/en/setup/service-agent/java-agent/README.md
+++ b/docs/en/setup/service-agent/java-agent/README.md
@@ -86,6 +86,7 @@ property key | Description | Default |
 `agent.force_reconnection_period `|Force reconnection period of grpc, based on grpc_channel_check_interval.|`1`|
 `agent.operation_name_threshold `|The operationName max length, setting this value > 190 is not recommended.|`150`|
 `agent.keep_tracing`|Keep tracing even the backend is not available if this value is `true`.|`false`|
+`agent.force_tls`|Force open TLS for gRPC channel if this value is `true`.|`true`|
 `osinfo.ipv4_list_size`| Limit the length of the ipv4 list size. |`10`|
 `collector.grpc_channel_check_interval`|grpc channel status check interval.|`30`|
 `collector.heartbeat_period`|agent heartbeat report period. Unit, second.|`30`|

--- a/docs/en/setup/service-agent/java-agent/README.md
+++ b/docs/en/setup/service-agent/java-agent/README.md
@@ -86,7 +86,7 @@ property key | Description | Default |
 `agent.force_reconnection_period `|Force reconnection period of grpc, based on grpc_channel_check_interval.|`1`|
 `agent.operation_name_threshold `|The operationName max length, setting this value > 190 is not recommended.|`150`|
 `agent.keep_tracing`|Keep tracing even the backend is not available if this value is `true`.|`false`|
-`agent.force_tls`|Force open TLS for gRPC channel if this value is `true`.|`true`|
+`agent.force_tls`|Force open TLS for gRPC channel if this value is `true`.|`false`|
 `osinfo.ipv4_list_size`| Limit the length of the ipv4 list size. |`10`|
 `collector.grpc_channel_check_interval`|grpc channel status check interval.|`30`|
 `collector.heartbeat_period`|agent heartbeat report period. Unit, second.|`30`|

--- a/docs/en/setup/service-agent/java-agent/TLS.md
+++ b/docs/en/setup/service-agent/java-agent/TLS.md
@@ -19,6 +19,10 @@ Only support **no mutual auth**.
 ### Agent config
 - Place `ca.crt` into `/ca` folder in agent package. Notice, `/ca` is not created in distribution, please create it by yourself.
 
-Agent open TLS automatically after the `/ca/ca.crt` file detected.
+Agent will open TLS upon the following config value. If it's true, TLS will be forced to use, and TLS works by default.
+At the same time, a trust manager will be constructed for TLS authentication once the `/ca/ca.crt` file detected.
+```
+agent.is_grpc_channel_tls_forced=${SW_AGENT_IS_GRPC_CHANNEL_TLS_FORCED:true}
+```
 
 o make sure can't access other ports out of region (VPC), such as firewall, proxy.

--- a/docs/en/setup/service-agent/java-agent/TLS.md
+++ b/docs/en/setup/service-agent/java-agent/TLS.md
@@ -22,7 +22,7 @@ Only support **no mutual auth**.
 - Agent open TLS automatically after the `/ca/ca.crt` file detected.
 - TLS with no CA mode could be activated by this setting.
 ```
-agent.force_tls=${SW_AGENT_FORCE_TLS:true}
+agent.force_tls=${SW_AGENT_FORCE_TLS:false}
 ```
 
 o make sure can't access other ports out of region (VPC), such as firewall, proxy.

--- a/docs/en/setup/service-agent/java-agent/TLS.md
+++ b/docs/en/setup/service-agent/java-agent/TLS.md
@@ -19,10 +19,10 @@ Only support **no mutual auth**.
 ### Agent config
 - Place `ca.crt` into `/ca` folder in agent package. Notice, `/ca` is not created in distribution, please create it by yourself.
 
-Agent will open TLS upon the following config value. If it's true, TLS will be forced to use, and TLS works by default.
-At the same time, a trust manager will be constructed for TLS authentication once the `/ca/ca.crt` file detected.
+- Agent open TLS automatically after the `/ca/ca.crt` file detected.
+- TLS with no CA mode could be activated by this setting.
 ```
-agent.is_grpc_channel_tls_forced=${SW_AGENT_IS_GRPC_CHANNEL_TLS_FORCED:true}
+agent.force_tls=${SW_AGENT_FORCE_TLS:true}
 ```
 
 o make sure can't access other ports out of region (VPC), such as firewall, proxy.

--- a/docs/en/setup/service-agent/java-agent/TLS.md
+++ b/docs/en/setup/service-agent/java-agent/TLS.md
@@ -24,5 +24,3 @@ Only support **no mutual auth**.
 ```
 agent.force_tls=${SW_AGENT_FORCE_TLS:false}
 ```
-
-o make sure can't access other ports out of region (VPC), such as firewall, proxy.

